CONFIG.md
# Takeout Configuration

## Music Files

Takeout will index all the objects in the S3 bucket to find music
files that end with supported file extensions. Since the actual files
are not individually opened to inspect tags, a specific file structure
is needed. MusicBrainz Picard is _highly_ recommended to manage and tag
music files.

The bucket structure should be:

	bucket/prefix/artist/album/track

where album can be:

	album
	album (year)

and track can be:

	1-track.mp3
	01-track.mp3
	1-01-track.mp3

### Examples

A single disc release by the Raconteurs in 2019:

	The Raconteurs/Help Us Stranger (2019)/01-Bored and Razed.flac

A multi disc release by Tubeway Army in 2019, track 1 from disc 1 and 2:

	Tubeway Army/Replicas - The First Recordings (2019)/1-01-You Are in My Vision (early version).flac
	Tubeway Army/Replicas - The First Recordings (2019)/2-01-Replicas (early version 2).flac

## Music Metadata

MusicBrainz APIs are used to obtain further information about artists
and albums found in the bucket. This is used to try to correct artist
and album names, and also to determine which tracks are singles or EPs
and their first release date.

Last.fm APIs are used to obtain popular track information for each
artist.

Fanart.tv APIs are used to obtain artist images.

Cover Art Archive APIs are used to obtain links to album covers.

## Movie Files

Takeout will index all the objects in the S3 bucket to find movie
files that end with supported file extensions. The actual files
are not individually opened to inspect tags or headers.

The bucket structure should be:

	bucket/prefix/path/movie

where movie can be:

    Title (year).mkv
    Title (year) - HD.mkv

### Examples

    Joker (2019) - HD.mkv
	Get Out (2017).mp4

## Movie Metadata

The Movie Database (TMDb) APIs are used to obtain all movie, cast and crew information.

## Configuration ##

The configuration file _takeout.ini_ is used to store required configuration to
setup and configure Takeout. Yaml and other formats supported by
[Viper](https://github.com/spf13/viper) can be used as well.


### Auth

* Auth.DB.Driver - Database driver (default sqlite3)
* Auth.DB.LogMode - Log SQL commands (default false)
* Auth.DB.Source - Name of database (default auth.db)
* Auth.MaxAge - Time until unused cookie expires (default 24h)
* Auth.SecureCookies - Use secure cookies (default true)

Note: only sqlite3 is supported

### Bucket

All of these are required and only a few have defaults.

* Bucket.Endpoint - Fully qualified domain name of the S3 bucket endpoint
* Bucket.Region - Bucket region
* Bucket.AccessKeyID - Access key for S3 bucket
* Bucket.SecretAccessKey - Secret key for S3 bucket
* Bucket.BucketName - Name of the S3 bucket
* Bucket.ObjectPrefix - Prefix of media within the S3 bucket
* Bucket.UseSSL - Use SSL with endpoint (default true)
* Bucket.URLExpiration - Time at with pre-signed URLs expire (default 72h)

Note: client is the AWS S3 Go client. Wasabi, Backblaze and Minio have been tested.

### Music Database

* Music.DB.Driver - Database driver (default sqlite3)
* Music.DB.Source - Name of database (default music.db)
* Music.DB.LogMore - Log SQL commands (default false)

Note: only sqlite3 is supported

### Music Settings

* Music.ArtistRadioBreadth - How many similar artists to use (default 10)
* Music.ArtistRadioDepth - How many similar artists tracks to include (default 3)
* Music.DeepLimit - How many deep tracks (default 50)
* Music.PopularLimit - How many popular tracks (default 50)
* Music.RadioLimit - How many radio tracks (default 25)
* Music.RadioSearchLimit - How many tracks to search for radio (default 1000)
* Music.Recent - How recent is recent (default 8760h = 1 year)
* Music.RecentLimit - How many recent (default 50)
* Music.SearchLimit - How many items to return (default 100)
* Music.SimilarArtistsLimit - Many many similar artists (default 10)
* Music.SimilarReleases - Duration for similar (default 8760h = 1 year)
* Music.SimilarRelesesLimit - Many many similar releases (default 10)
* Music.SinglesLimit - How many singles (default 50)

### Server

* Server.Listen - Address and port to listen on (default 127.0.0.1:3000)
* Server.WebDir - Directory with web static files and templates

Note: use Nginx or other frontend with TLS (and Let's Encrypt)

### Search Index

* Search.BleveDir - Directory used to store index (default .)

### HTTP Client

* Client.UseCache - Enable or disable http caching (default false)
* Client.MaxAge - Age in seconds to use cached responses (default 30 days)
* Client.CacheDir - Directory to store cached responses (default .httpcache)

Note: this client is used for all API calls except Last.fm

## Examples

### Required fields in ini format

	[bucket]
	endpoint = "s3.us-west-1.wasabisys.com"
	region = "us-west-1"
	accessKeyID = "<insert access key here>"
	secretAccessKey = "<insert secret key here>"
	bucketName = "MyBucket"
	objectPrefix = "MyMedia"


config/config.go
// Copyright (C) 2020 The Takeout Authors.
//
// This file is part of Takeout.
//
// Takeout is free software: you can redistribute it and/or modify it under the
// terms of the GNU Affero General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option)
// any later version.
//
// Takeout is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
// more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Takeout.  If not, see <https://www.gnu.org/licenses/>.

package config

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/defsub/takeout"
	"github.com/spf13/viper"
)

const (
	MediaMusic = "music"
	MediaVideo = "video"
)

type BucketConfig struct {
	Endpoint        string
	Region          string
	AccessKeyID     string
	SecretAccessKey string
	BucketName      string
	ObjectPrefix    string
	UseSSL          bool
	URLExpiration   time.Duration
	Media           string
}

type DatabaseConfig struct {
	Driver  string
	Source  string
	LogMode bool
}

type MusicConfig struct {
	ArtistFile           string
	ArtistRadioBreadth   int
	ArtistRadioDepth     int
	DB                   DatabaseConfig
	DeepLimit            int
	PopularLimit         int
	RadioGenres          []string
	RadioLimit           int
	RadioOther           map[string]string
	RadioSearchLimit     int
	RadioSeries          []string
	Recent               time.Duration
	RecentLimit          int
	ReleaseCountries     []string
	ReleaseFile          string
	SearchLimit          int
	SimilarArtistsLimit  int
	SimilarReleases      time.Duration
	SimilarReleasesLimit int
	SinglesLimit         int
	artistMap            map[string]string
}

type VideoConfig struct {
	DB               DatabaseConfig
	ReleaseCountries []string
	CastLimit        int
	CrewJobs         []string
	Recent           time.Duration
	RecentLimit      int
	SearchLimit      int
}

type LastFMAPIConfig struct {
	Key    string
	Secret string
}

type FanartAPIConfig struct {
	ProjectKey  string
	PersonalKey string
}

type TMDBAPIConfig struct {
	Key string
}

type AuthConfig struct {
	DB            DatabaseConfig
	MaxAge        time.Duration
	SecureCookies bool
}

type SearchConfig struct {
	BleveDir string
}

type ServerConfig struct {
	Listen string
	WebDir string
	URL    string
}

type ClientConfig struct {
	CacheDir  string
	MaxAge    int
	UseCache  bool
	UserAgent string
}

type Config struct {
	Auth    AuthConfig
	Buckets []BucketConfig
	Client  ClientConfig
	DataDir string
	Fanart  FanartAPIConfig
	LastFM  LastFMAPIConfig
	Music   MusicConfig
	TMDB    TMDBAPIConfig
	Search  SearchConfig
	Server  ServerConfig
	Video   VideoConfig
}

func (mc *MusicConfig) UserArtistID(name string) (string, bool) {
	mbid, ok := mc.artistMap[name]
	return mbid, ok
}

func readJsonStringMap(file string, m *map[string]string) (err error) {
	data, err := ioutil.ReadFile(file)
	if err != nil {
		return
	}
	err = json.Unmarshal([]byte(data), m)
	return
}

func (mc *MusicConfig) readMaps() {
	if mc.ArtistFile != "" {
		readJsonStringMap(mc.ArtistFile, &mc.artistMap)
	}
}

func configDefaults(v *viper.Viper) {
	v.SetDefault("Auth.DB.Driver", "sqlite3")
	v.SetDefault("Auth.DB.LogMode", "false")
	v.SetDefault("Auth.DB.Source", "auth.db")
	v.SetDefault("Auth.MaxAge", "24h")
	v.SetDefault("Auth.SecureCookies", "true")

	// TODO apply as default
	// v.SetDefault("Bucket.URLExpiration", "72h")
	// v.SetDefault("Bucket.UseSSL", "true")

	v.SetDefault("Client.CacheDir", ".httpcache")
	v.SetDefault("Client.MaxAge", 86400*30) // 30 days in seconds
	v.SetDefault("Client.UseCache", "false")
	v.SetDefault("Client.UserAgent", userAgent())

	v.SetDefault("DataDir", ".")

	v.SetDefault("Fanart.ProjectKey", "93ede276ba6208318031727060b697c8")

	v.SetDefault("LastFM.Key", "77033164cfcda2acc4c58681dcba3cf8")
	v.SetDefault("LastFM.Secret", "8f43410e8e81c33d4542738ee84dc39b")

	v.SetDefault("Music.ArtistRadioBreadth", "10")
	v.SetDefault("Music.ArtistRadioDepth", "3")
	v.SetDefault("Music.DeepLimit", "50")
	v.SetDefault("Music.PopularLimit", "50")
	v.SetDefault("Music.RadioLimit", "25")
	v.SetDefault("Music.RadioSearchLimit", "1000")
	v.SetDefault("Music.Recent", "8760h") // 1 year
	v.SetDefault("Music.RecentLimit", "50")
	v.SetDefault("Music.SearchLimit", "100")
	v.SetDefault("Music.SimilarArtistsLimit", "10")
	v.SetDefault("Music.SimilarReleases", "8760h") // +/- 1 year
	v.SetDefault("Music.SimilarReleasesLimit", "10")
	v.SetDefault("Music.SinglesLimit", "50")

	// see https://wiki.musicbrainz.org/Release_Country
	// v.SetDefault("Music.ReleaseCountries", []string{
	// 	"US", // United States
	// 	"XW", // Worldwide
	// 	"XE", // Europe
	// })

	v.SetDefault("Music.DB.Driver", "sqlite3")
	v.SetDefault("Music.DB.Source", "music.db")
	v.SetDefault("Music.DB.LogMode", "false")

	v.SetDefault("TMDB.Key", "903a776b0638da68e9ade38ff538e1d3")

	v.SetDefault("Search.BleveDir", ".")

	v.SetDefault("Server.Listen", "127.0.0.1:3000")
	v.SetDefault("Server.WebDir", "web")
	v.SetDefault("Server.URL", "https://example.com") // w/o trailing slash

	v.SetDefault("Video.DB.Driver", "sqlite3")
	v.SetDefault("Video.DB.Source", "video.db")
	v.SetDefault("Video.DB.LogMode", "true")
	v.SetDefault("Video.ReleaseCountries", []string{
		"US",
	})
	v.SetDefault("Video.CastLimit", "25")
	v.SetDefault("Video.CrewJobs", []string{
		"Director",
		"Executive Producer",
		"Novel",
		"Producer",
		"Screenplay",
		"Story",
	})
	v.SetDefault("Video.Recent", "8760h") // 1 year
	v.SetDefault("Video.RecentLimit", "50")
	v.SetDefault("Video.SearchLimit", "100")

	// see https://musicbrainz.org/search (series)
	v.SetDefault("Music.RadioSeries", []string{
		"The Rolling Stone Magazine's 500 Greatest Songs of All Time",
	})

	v.SetDefault("Music.RadioOther", map[string]string{
		"Series Hits": "+series:*",
		"Top Hits":    "+popularity:1",
		"Top 3 Hits":  "+popularity:<4",
		"Top 5 Hits":  "+popularity:<6",
		"Top 10 Hits": "+popularity:<11",
		"Covers":      "+type:cover",
		"Live Hits":   "+type:live +popularity:<3",
	})
}

func userAgent() string {
	return takeout.AppName + "/" + takeout.Version + " ( " + takeout.Contact + " ) "
}

func readConfig(v *viper.Viper) (*Config, error) {
	var config Config
	var pathRegexp = regexp.MustCompile(`(file|dir|source)$`)
	err := v.ReadInConfig()
	dir := filepath.Dir(v.ConfigFileUsed())
	for _, k := range v.AllKeys() {
		if pathRegexp.MatchString(k) {
			val := v.Get(k)
			if strings.HasPrefix(val.(string), "/") == false {
				val = fmt.Sprintf("%s/%s", dir, val.(string))
				v.Set(k, val)
			}
		}
	}
	if err == nil {
		err = v.Unmarshal(&config)
		config.Music.readMaps()
	}
	return &config, err
}

func TestConfig() (*Config, error) {
	testDir := os.Getenv("TEST_CONFIG")
	if testDir == "" {
		return nil, errors.New("missing test config")
	}
	v := viper.New()
	configDefaults(v)
	v.SetConfigFile(filepath.Join(testDir, "test.yaml"))
	v.SetDefault("Music.DB.Source", filepath.Join(testDir, "music.db"))
	v.SetDefault("Auth.DB.Source", filepath.Join(testDir, "auth.db"))
	return readConfig(v)
}

var configFile, configPath, configName string

func SetConfigFile(path string) {
	configFile = path
}

func AddConfigPath(path string) {
	configPath = path
}

func SetConfigName(name string) {
	configName = name
}

func GetConfig() (*Config, error) {
	v := viper.New()
	if configFile != "" {
		v.SetConfigFile(configFile)
	}
	if configPath != "" {
		v.AddConfigPath(configPath)
	}
	if configName != "" {
		v.SetConfigName(configName)
	}
	configDefaults(v)
	return readConfig(v)
}

func LoadConfig(dir string) (*Config, error) {
	v := viper.New()
	v.AddConfigPath(dir)
	configDefaults(v)
	return readConfig(v)
}
